<a href="https://colab.research.google.com/github/genairsh/GenAI_Level3_Assignment/blob/main/Problem1_EmpProductivityScore.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
#step 1: import the libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [4]:
#step 2: reading the data and splitting it to input and output
dataset = pd.read_csv('https://raw.githubusercontent.com/genairsh/GenAI_Level3_Assignment/main/datasets/employee_productivity.csv')
inputx = dataset.iloc[:, :-1].values
outputy = dataset.iloc[:, -1].values
print(outputy)

[160 296 226 305 263 266 309 235 174 226 189 171 290 204 281 200 221 196
 281 325 297 302 255 305 261 160 182 355 153 275 206 330 281 266 217 219
 101 235 173 284 146 302 129 193 232 249 274 241 142 115]


In [5]:
#step 3: select one thirds of the data for testing and two thirds for training.
input_train, input_test, output_train, output_test = train_test_split(
                                                    inputx, outputy,
                                                    test_size = 1/4, random_state = 42)

In [6]:
#step 4: selecting the Linear Regression model
model = LinearRegression()
print("\nThe parameters of the model are\n\n",model.get_params())
model.fit(input_train, output_train)
print("\nThe model we are using is ",model)


The parameters of the model are

 {'copy_X': True, 'fit_intercept': True, 'n_jobs': None, 'positive': False}

The model we are using is  LinearRegression()


In [7]:
#step 5: testing or model prediction using testinput
predicted_output = model.predict(input_test)


In [8]:
# Step 6: Evaluate Model
print("Model Evaluation")
print("----------------------------")
print("R2 Score :", r2_score(output_test, predicted_output))
print("MAE      :", mean_absolute_error(output_test, predicted_output))
print("MSE      :", mean_squared_error(output_test, predicted_output))
print("RMSE     :", np.sqrt(mean_squared_error(output_test, predicted_output)))


Model Evaluation
----------------------------
R2 Score : 0.8782808448586914
MAE      : 14.511699621741123
MSE      : 345.6535913338274
RMSE     : 18.591761383307055


In [9]:
# -----------------------------
# Step 7: Regression Equation
# -----------------------------
print("\nIntercept:", model.intercept_)

print("\nCoefficients")
##Select dataset all columns(i.e features) except output column(i.e last column)
columns = dataset.columns[:-1]

for name, coef in zip(columns, model.coef_):
    print(name, ":", coef)


Intercept: 29.577633900149124

Coefficients
Experience (yrs) : 5.3685547579555415
Training Hours : 0.46986993853163117
Working Hours : 1.475963620832422
Projects : 9.091041858349026


In [10]:
# -----------------------------
# Step 8: Test using one training record
# -----------------------------
sample = input_test[0].reshape(1, -1)

predicted = model.predict(sample)

print("\nTesting with one training record")
print("--------------------------------")
print("Input :", sample)
print("Actual Productivity :", output_train[0])
print("Predicted Productivity :", predicted[0])


Testing with one training record
--------------------------------
Input : [[ 6 24 39  9]]
Actual Productivity : 174
Predicted Productivity : 212.44779891024723


In [11]:
# -----------------------------
# Step 9: Predict New Employee
# -----------------------------
# Experience, Training Hours,
# Working Hours, Projects

new_employee = [[10, 94, 48, 12]]

prediction = model.predict(new_employee)

print("\nPredicted Productivity for New Employee:")
print(prediction[0])


Predicted Productivity for New Employee:
307.36971180182246


**Interpretation**

**Which factor most strongly impacts productivity?**
The number of Projects has the strongest positive impact on productivity because it has the largest regression coefficient (9.09).
This means that, keeping all other factors constant, completing one additional project increases the predicted productivity score by about 9.09 points.

**How does training affect productivity? **The coefficient for Training Hours is 0.47, which is positive.
This means that increasing training hours increases productivity. Specifically, for every additional hour of training, the productivity score is expected to increase by approximately 0.47 points, assuming all other factors remain unchanged.
Training has a positive effect, but its impact is smaller than Experience and Projects.


**Should the company increase training hours or working hours?** Based on the model:

Training Hours coefficient = 0.47
Working Hours coefficient = 1.48

Since the coefficient for Working Hours is larger, the model predicts that increasing working hours has a greater immediate impact on productivity than increasing training hours.
However, from a practical business perspective, increasing working hours indefinitely may cause employee fatigue and burnout. Therefore, the company should increase working hours only within healthy limits while continuing to invest in training for long-term employee development.

**What happens if Working Hours increase beyond optimal limits?** This Linear Regression model assumes a linear relationship and predicts that productivity continues to increase by about 1.48 points for each additional working hour.
However, in reality, employees may become tired or stressed if they work too many hours. Beyond an optimal limit, productivity may stop increasing or even decrease.

Since Linear Regression cannot model this nonlinear effect, a more advanced model (such as Polynomial Regression) would be needed to capture it.


**Can productivity ever decrease with more experience?** According to this model, No.

The Experience coefficient is +5.37, which means productivity is predicted to increase by approximately 5.37 points for each additional year of experience, assuming other variables remain constant.

If the coefficient had been negative, the model would have predicted decreasing productivity with increasing experience.

**How would you detect overfitting in this model?**
Overfitting can be detected by comparing the model's performance on the training and testing datasets.

If the training accuracy (R² score) is much higher than the testing accuracy, the model may be overfitting the training data.

Other methods include:
Comparing training and testing R² scores.
Using cross-validation.
Testing the model on new, unseen data.

Since this model achieved a good testing R² score of 0.878, there is no strong evidence of overfitting based on the available information.

**Suggest one new feature to improve prediction accuracy.**
Job Satisfaction Score, as it can significantly influence employee productivity.
